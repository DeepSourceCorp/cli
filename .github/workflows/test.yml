name: Unit Tests

on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      NODE_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}
      ZEAL_SECRET: ${{secrets.GITHUB_TOKEN}}

    steps:
      - name: Checkout Bifrost
        uses: actions/checkout@v2

      - name: Setting Up Node
        uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          always-auth: true
          registry-url: https://npm.pkg.github.com/
          scope: '@deepsourcelabs'
          cache: 'yarn'

      - name: Install packages
        run: yarn

      - name: Run Test
        run: yarn test --coverage=true --coverageReporters=cobertura

      - name: Install DeepSource CLI
        run: curl https://deepsource.io/cli | sh

      - name: Set DEEPSOURCE_DSN
        run: |
          ./bin/deepsource report --analyzer test-coverage --key javascript --value-file ./coverage/cobertura-coverage.xml
        env:
          DEEPSOURCE_DSN: ${{ secrets.BIFROST_DSN }}
