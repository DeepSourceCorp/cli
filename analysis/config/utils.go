package config

import (
	"log"
	"os"
	"path"
	"path/filepath"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
)

/* Walks the `CODE_PATH` directory and returns all the files other than the ones present
 * in the .git folder for analysis in the form of a string array */
func readAllFiles(codePath string) ([]string, error) {
	fileCount := 0

	allFiles := make([]string, 0)

	err := filepath.Walk(codePath,
		func(path string, fileInfo os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			fileCount++

			/* Check the following before appending to the list of files:
			 * - Should not be a directory
			 * - The walked file should not be present in .git folder
			 * TODO: Check why fileCount != 1 condition is used here */
			if !fileInfo.IsDir() || !strings.HasPrefix(path, filepath.Join(codePath, ".git")) && fileCount != 1 {
				allFiles = append(allFiles, path)
			}
			return nil
		})
	if err != nil {
		return allFiles, err
	}
	return allFiles, nil
}

// Returns the slice of files matching certain glob patterns
func (r *AnalysisRun) getMatchingFiles(patterns []string) ([]string, error) {
	matchedFiles := make([]string, 0)

	// Return all the files if no exclude_patterns are configured
	if len(patterns) == 0 {
		return matchedFiles, nil
	}

	for _, file := range r.AnalysisFiles {
		for i := range patterns {
			match, err := doublestar.Match(path.Join(r.CodePath, patterns[i]), file)
			if err != nil {
				log.Printf("Failed to match the file %s with pattern %s\n", file, patterns[i])
			}
			if match {
				matchedFiles = append(matchedFiles, file)
				break
			}
		}
	}
	return matchedFiles, nil
}

// Filters the analysis files and removes the files matching the exclude_patterns from them
// TODO: Improve the logic here
func (r *AnalysisRun) filterAnalysisFiles() {
	filteredFiles := []string{}
	for _, file := range r.AnalysisFiles {
		for _, excludedFile := range r.ExcludedFiles {
			if file == excludedFile {
				break
			}
			filteredFiles = append(filteredFiles, file)
		}
	}
	r.AnalysisFiles = filteredFiles
}
