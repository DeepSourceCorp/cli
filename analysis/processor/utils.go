package processor

import (
	"sort"

	"github.com/deepsourcelabs/cli/types"
)

// sortIssuesByFile sorts the issues in an alphabetical order according to the filenames
// where they got reported.
func (p *ProcessAnalysisResults) sortIssuesByFile() {
	sort.Slice(p.AnalysisResult.Issues, func(i, j int) bool {
		return p.AnalysisResult.Issues[i].Location.Path < p.AnalysisResult.Issues[j].Location.Path
	})
}

// Prepare a map with unique filenames as key and the issue range for each file as value
// This is done to ensure fewer loops when processing issues.
type IssueRange struct {
	Filename   string // The file which has the issues
	BeginIndex int    // Array index in report.Issues where the particular issue starts
	EndIndex   int    // Array index in report.Issues where the particular issue ends
}

// GenerateIssueRangeSlice generates an array containing the issue ranges with respect to files
// that helps us to go through them and map them to the files where they got reported instead
// of opening the file for each of them. The generated index looks like this:

// [{analyzer.go 0 0} {autofix_patch.go 1 1} {difftool.go 2 2} {patch.patch 3 5} {proc_skip_cq_test.go 6 6}]

// Here, the first field if filename and the second and third fields are the index range in which the issues reported
// in these files lie in the sorted AnalyzerReport slice.
func createIssueFileRange(report types.AnalysisResult) []IssueRange {
	fileCount := 0 // for 1 file, 0 based indexing
	issuesRange := []IssueRange{}
	prevFilename := report.Issues[0].Location.Path

	issueRange := IssueRange{
		BeginIndex: 0,
		EndIndex:   len(report.Issues) - 1,
		Filename:   prevFilename,
	}
	issuesRange = append(issuesRange, issueRange)

	// Iterating over the issues and creating an array containing issues with index data about
	// the files in which those issues are present
	for i := 1; i < len(report.Issues); i++ {
		issue := report.Issues[i]
		currentFilename := issue.Location.Path

		// TODO: Check when this condition is implied
		if issue.Location.Position.End.Line == -1 {
			issue.Location.Position.End.Line = issue.Location.Position.Begin.Line
		}

		if currentFilename != prevFilename {
			fileCount++

			issueRange = issuesRange[fileCount-1]
			issueRange.EndIndex = i - 1
			issuesRange[fileCount-1] = issueRange

			// Create for the new file
			issueRange = IssueRange{
				Filename:   currentFilename,
				BeginIndex: i,
				EndIndex:   len(report.Issues) - 1,
			}

			issuesRange = append(issuesRange, issueRange)
			prevFilename = currentFilename
		}
	}
	return issuesRange
}

// formatLSPResultsToDefault converts the LSP based analysis results into the default format supported by DeepSource.
func (p *ProcessAnalysisResults) formatLSPResultsToDefault() {
	p.AnalysisResult.IsPassed = p.Report.IsPassed
	p.AnalysisResult.Metrics = append(p.Report.Metrics, p.AnalysisResult.Metrics...)
	p.AnalysisResult.Errors = append(p.Report.Errors, p.AnalysisResult.Errors...)

	// Appending the issues to the default format of Analysis report
	for _, issue := range p.Report.Issues {
		analysisIssue := types.Issue{
			IssueCode: issue.Code,
			IssueText: issue.Message,
			Location: types.Location{
				Path: issue.RelatedInformation[0].Location.URI,
				Position: types.Position{
					Begin: types.Coordinate{
						Line:   issue.Range.Start.Line,
						Column: issue.Range.Start.Character,
					},
					End: types.Coordinate{
						Line:   issue.Range.End.Line,
						Column: issue.Range.End.Character,
					},
				},
			},
		}
		p.AnalysisResult.Issues = append(p.AnalysisResult.Issues, analysisIssue)
	}
}
