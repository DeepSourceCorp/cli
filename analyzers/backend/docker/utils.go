package docker

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

type ErrorLine struct {
	Error       string      `json:"error"`
	ErrorDetail ErrorDetail `json:"errorDetail"`
}

type ErrorDetail struct {
	Message string `json:"message"`
}

type DockerBuildResponse struct {
	Stream string `json:"stream"`
}

func CheckBuildResponse(rd io.Reader, showAllLogs bool) error {
	var lastLine []byte
	count := 0
	var currentStream string

	scanner := bufio.NewScanner(rd)
	for scanner.Scan() {
		lastLine = scanner.Bytes()
		d := &DockerBuildResponse{}
		err := json.Unmarshal(lastLine, d)
		if err != nil {
			return err
		}
		if d.Stream == "" || d.Stream == "\n" || strings.Contains(d.Stream, "--->") || strings.TrimSuffix(d.Stream, "\n") == currentStream {
			continue
		}
		currentStream = strings.TrimSuffix(d.Stream, "\n")
		if showAllLogs {
			fmt.Println(currentStream)
		}
		count++
	}

	errLine := &ErrorLine{}
	json.Unmarshal([]byte(lastLine), errLine)
	if errLine.Error != "" {
		return errors.New(errLine.Error)
	}
	return scanner.Err()
}


