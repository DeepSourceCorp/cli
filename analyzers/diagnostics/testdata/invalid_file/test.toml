name = "2do-Checker"
