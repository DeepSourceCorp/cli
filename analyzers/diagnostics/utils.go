package diagnostics

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/morikuni/aec"
)

// fieldRegexp is a regular expression for checking the current line is a field or not.
var fieldRegexp = `\s*(?P<field>.+)\s*=\s*`

// readFileContent reads the file and returns its content.
func readFileContent(filename string) (string, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

// prepareCodeFrame prepares a code frame using the file content. The code frame is meant to be displayed on the console while reporting diagnostics.
// NOTE: lineNum always starts from 0.
func prepareCodeFrame(lineNum int, lines []string) string {
	frame := ""

	if lineNum <= 1 {
		// Case 1: When the line is near the top of the file.
		// Generate a frame with the current and next line only.
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])

	} else if lineNum >= (len(lines) - 1) {
		// Case 2: When the line is near the bottom of the file.
		// Generate a frame with the current line only.
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum, lines[lineNum-1]))
	} else {
		// Case 3: When the line is in between the top and the bottom.
		// Generate a frame with the the previous, current and the next line.
		frame += fmt.Sprintf("  %d | %s\n", lineNum, lines[lineNum-1])
		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])
	}

	return frame
}

// checkField checks if the line contains the field.
func checkField(line, field string) bool {
	// We use fieldRegexp for checking if the current line is a field or not.
	// If it's not a field, for example, comments for explanation, etc., then checkField should return false.
	exp := regexp.MustCompile(fieldRegexp)

	// Get index for the "field" named group.
	matches := exp.FindStringSubmatch(line)
	groupIndex := exp.SubexpIndex("field")

	// Capture match. We proceed only if the groupIndex isn't -1 (group not found), and if the number of matches aren't zero.
	// Once done, we populate match with the match found using the regular expression.
	match := ""
	if len(matches) != 0 && groupIndex != -1 {
		match = matches[groupIndex]
	}
	match = strings.TrimSpace(match)

	// Return true if the match contains the field.
	return strings.Contains(match, field)
}
