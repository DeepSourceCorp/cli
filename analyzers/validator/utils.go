package validator

import (
	"errors"
	"fmt"
	"path"
	"reflect"
	"regexp"
	"strings"

	validate "github.com/go-playground/validator/v10"
	"github.com/pelletier/go-toml/v2"
)

// codeframeFieldRegexp is a regular expression used for matching a TOML field present in the codeframe generated by go-toml.
var codeframeFieldRegexp = `\d\s*[|].(?P<field>\w+)\s*=\s*\S*`

// Returns the list of required fields from the error message returned by the `go-playground/validator` library
func getMissingRequiredFields(err error, config interface{}) []string {
	missingRequiredFields := []string{}
	errs := err.(validate.ValidationErrors)
	for _, err := range errs {
		if err.Tag() == "required" {
			c := reflect.ValueOf(config)
			for i := 0; i < c.Type().NumField(); i++ {
				if err.Field() == c.Type().Field(i).Name {
					missingRequiredFields = append(missingRequiredFields, c.Type().Field(i).Tag.Get("toml"))
				}
			}
		}
	}
	return missingRequiredFields
}

// Handle decoding errors reported by go-toml
func handleTOMLDecodeErrors(err error, filePath string) *ValidationFailure {
	var usefulResponse, expectedType, receivedType, fieldName, decodeErrorMessage string

	// We check for go-toml's StrictMissingError, as the decoder returns a StrictMissingError when there exists a mismatch between the destination and the extracted fields.
	// go-toml doesn't return the field name associated with the error, however, StrictMissingError returns a human-readable codeframe, which can be used to extract the error-causing fields.
	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#example-Decoder.DisallowUnknownFields

	var strictMissingError *toml.StrictMissingError
	if errors.As(err, &strictMissingError) {
		// Extract human-readable codeframe.
		codeframe := strictMissingError.String()

		// Retrieve field names from the codeframe.
		fields := parseCodeframe(codeframe)

		// Construct errors using the extracted field names.
		var errors []ErrorMeta
		for _, field := range fields {
			errMeta := ErrorMeta{
				Level:   DecodeErr,
				Field:   field,
				Message: fmt.Sprintf("Invalid field found: %s", field),
			}
			errors = append(errors, errMeta)
		}

		// Construct a validation failure.
		validationError := ValidationFailure{
			File:   filePath,
			Errors: errors,
		}

		return &validationError
	}

	// Get the DecodeError exported by go-toml
	// Ref: https://pkg.go.dev/github.com/pelletier/go-toml/v2#DecodeError
	var decodeErr *toml.DecodeError
	if !errors.As(err, &decodeErr) {
		decodeErrorMessage = err.Error()

		// Handle strict mode error when some alien fields are added in the user configured TOML
		if strings.HasPrefix(err.Error(), "strict mode") {
			decodeErrorMessage = fmt.Sprintf("Failed to parse %s. Invalid fields detected.", path.Base(filePath))
		}
		validationError := ValidationFailure{
			File: filePath,
			Errors: []ErrorMeta{
				{
					Level:   DecodeErr,
					Field:   "",
					Message: decodeErrorMessage,
				},
			},
		}
		return &validationError
	}

	/* =================================================
	 * Extract the data about the decoding failure and return
	 * a validation failure response
	 * ================================================= */

	errorMessage := decodeErr.Error()
	// Error case 1: `toml: cannot decode TOML integer into struct field types.AnalyzerTOML.Name of type string"`
	if strings.HasPrefix(errorMessage, "toml: cannot decode TOML") {

		usefulResponse = strings.TrimPrefix(errorMessage, "toml: cannot decode TOML ")
		responseArray := strings.Split(usefulResponse, " ")

		expectedType = responseArray[len(responseArray)-1]
		receivedType = responseArray[0]
		fieldData := responseArray[len(responseArray)-4]
		index := strings.LastIndex(fieldData, ".")
		fieldName = strings.ToLower(fieldData[index:])
		// Framing the decoding failure error message
		decodeErrorMessage = fmt.Sprintf("expected the field \"%s\" of type %s. Got %s.", fieldName, expectedType, receivedType)

	} else if strings.HasPrefix(errorMessage, "toml: cannot store TOML") {

		// Error case 2: `toml: cannot store TOML string into a Go slice`
		usefulResponse = strings.TrimPrefix(errorMessage, "toml: cannot store TOML ")
		responseArray := strings.Split(usefulResponse, " ")

		expectedType = responseArray[len(responseArray)-1]
		receivedType = responseArray[0]
		decodeErrorMessage = fmt.Sprintf("expected type for one of the fields : %s. Received: %s.", expectedType, receivedType)
	} else {
		decodeErrorMessage = errorMessage
		fieldName = ""
	}

	validationError := ValidationFailure{
		File: filePath,
		Errors: []ErrorMeta{
			{
				Level:   DecodeErr,
				Field:   fieldName,
				Message: decodeErrorMessage,
			},
		},
	}
	return &validationError
}

// parseCodeframe parses the codeframe generated by go-toml, and extracts field names which are associated with the error.
func parseCodeframe(codeframe string) []string {
	var fields []string
	lines := strings.Split(codeframe, "\n")

	// Compile a regular expression for matching fields.
	exp := regexp.MustCompile(codeframeFieldRegexp)

	// This is added as a safety check. pelletier/go-toml/v2 doesn't return codeframes lesser than 2 lines.
	if len(lines) < 2 {
		return []string{""}
	}

	for i := 1; i < len(lines); i++ {
		// Retrieve current and previous line from the codeframe.
		currLine := lines[i]
		prevLine := lines[i-1]

		// Check if the current line contains the "missing field" message.
		if strings.Contains(currLine, "missing field") {
			// If yes, use the codeframeFieldRegexp regular expression for matching the previous line to retrieve the field name.
			matches := exp.FindStringSubmatch(prevLine)

			// Get index for the "field" named group.
			groupIndex := exp.SubexpIndex("field")

			// Capture match. We proceed only if the groupIndex isn't -1 (group not found), and if the number of matches aren't zero.
			match := ""
			if len(matches) != 0 && groupIndex != -1 {
				match = matches[groupIndex]
			}

			// Add the field name to fields.
			field := strings.TrimSpace(match)
			fields = append(fields, field)
		}
	}

	return fields
}
