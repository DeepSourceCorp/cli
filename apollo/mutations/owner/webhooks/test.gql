mutation testWebhook($webhookId: ID!) {
  testWebhook(input: { webhookId: $webhookId }) {
    ok
  }
}
