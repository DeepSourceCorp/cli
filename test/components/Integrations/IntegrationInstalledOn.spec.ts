import '@testing-library/jest-dom'
import { render } from '@testing-library/vue'

import IntegrationInstalledOn from '~/components/Integrations/IntegrationInstalledOn.vue'
import { cartesian, generateStringProps } from '~/test/utils'

describe('[[ IntegrationInstalledOn ]]', () => {
  test('renders the integration installed on section', () => {
    const enabledOnOptions = generateStringProps('installedOn', ['deepsourcelabs'])

    cartesian(enabledOnOptions).forEach((props) => {
      const { html } = render(IntegrationInstalledOn, {
        props,
        stubs: {
          Notice: true
        }
      })

      expect(html()).toMatchSnapshot(JSON.stringify(props))
    })
  })
})
