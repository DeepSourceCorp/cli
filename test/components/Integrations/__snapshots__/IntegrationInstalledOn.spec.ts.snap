// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[ IntegrationInstalledOn ]] renders the integration installed on section: {"installedOn":"deepsourcelabs"} 1`] = `
"<notice-stub class=\\"h-8\\">
  <p class=\\"text-xs truncate\\">
    Installed on <span class=\\"font-medium text-vanilla-100\\">deepsourcelabs</span></p>
</notice-stub>"
`;

exports[`[[ IntegrationInstalledOn ]] renders the integration installed on section: {"installedOn":null} 1`] = `
"<notice-stub class=\\"h-8\\">
  <p class=\\"text-xs truncate\\">
    Installed
    <!---->
    <!---->
  </p>
</notice-stub>"
`;

exports[`[[ IntegrationInstalledOn ]] renders the integration installed on section: {} 1`] = `
"<notice-stub class=\\"h-8\\">
  <p class=\\"text-xs truncate\\">
    Installed
    <!---->
    <!---->
  </p>
</notice-stub>"
`;
